python/ray/rllib/test/test_common_policy_evaluator.py

```diff
@@ -30,43 +83,111 @@ def testBasic(self):
             self.assertIn(key, batch)
         self.assertGreater(batch["advantages"][0], 1)
 
-    def testPackEpisodes(self):
-        for batch_size in [1, 10, 100, 1000]:
-            ev = CommonPolicyEvaluator(
-                env_creator=lambda _: gym.make("CartPole-v0"),
-                policy_graph=MockPolicyGraph,
-                batch_steps=batch_size,
-                batch_mode="pack_episodes")
-            batch = ev.sample()
-            self.assertEqual(batch.count, batch_size)
+    def testQueryEvaluators(self):
+        register_env("test", lambda _: gym.make("CartPole-v0"))
+        pg = PGAgent(env="test", config={"num_workers": 2, "batch_size": 5})
+        results = pg.optimizer.foreach_evaluator(lambda ev: ev.batch_steps)
+        results2 = pg.optimizer.foreach_evaluator_with_index(
+            lambda ev, i: (i, ev.batch_steps))
+        self.assertEqual(results, [5, 5, 5])
+        self.assertEqual(results2, [(0, 5), (1, 5), (2, 5)])
 
-    def testTruncateEpisodes(self):
+    def testMetrics(self):
+        ev = CommonPolicyEvaluator(
+            env_creator=lambda _: MockEnv(episode_length=10),
+            policy_graph=MockPolicyGraph, batch_mode="complete_episodes")
+        remote_ev = CommonPolicyEvaluator.as_remote().remote(
+            env_creator=lambda _: MockEnv(episode_length=10),
+            policy_graph=MockPolicyGraph, batch_mode="complete_episodes")
+        ev.sample()
+        ray.get(remote_ev.sample.remote())
+        result = collect_metrics(ev, [remote_ev])
+        self.assertEqual(result.episodes_total, 20)
+        self.assertEqual(result.episode_reward_mean, 10)
+
+    def testAsync(self):
         ev = CommonPolicyEvaluator(
             env_creator=lambda _: gym.make("CartPole-v0"),
+            sample_async=True,
+            policy_graph=MockPolicyGraph)
+        batch = ev.sample()
+        for key in ["obs", "actions", "rewards", "dones", "advantages"]:
+            self.assertIn(key, batch)
+        self.assertGreater(batch["advantages"][0], 1)
+
+    def testAutoConcat(self):
+        ev = CommonPolicyEvaluator(
+            env_creator=lambda _: MockEnv(episode_length=40),
             policy_graph=MockPolicyGraph,
-            batch_steps=2,
-            batch_mode="truncate_episodes")
+            sample_async=True,
+            batch_steps=10,
+            batch_mode="truncate_episodes",
+            observation_filter="ConcurrentMeanStdFilter")
+        time.sleep(2)
```

Comment:
iirc using `time` in these tests often lead to weird results...

Reply:
Yeah let's see if it's flaky.